src/bisect.cpp

```diff
@@ -132,23 +153,104 @@ std::vector<long> bisect_multilevel(bulk::world& world,
         }
 
 
-        std::map<std::vector<long>, int> edge_counts;
-        std::map<std::vector<long>, long> ids;
-        std::unordered_set<long> remove_nets;
+        if (simplify_mode == "complete") {
+            if (world.rank() == 0) {
+                world.log("Size before par simplify: %d", HC_list[nc_par].global_number_nets());
+            }
 
-        for (auto n : HC_list[nc_par].nets()) {
-            edge_counts[n.vertices()] += n.cost();
-            ids[n.vertices()] = n.id();
-        }
+            // Communicate globally which net ids belong to the local hypergraphs
+            auto netidowner_queue = bulk::queue<int, long>(world);
+            for (auto n : HC_list[nc_par].nets()) {
+                if (n.size() != n.global_size()) {
+                    for (int t = 0u; t < world.active_processors(); t++) {
+                        if (world.rank() != t) {
+                            netidowner_queue(t).send(world.rank(), n.id());
+                        }
+                    }
+                }
+            }
+            world.sync();
+
+            // Communicate pins targeted only to relevant processors
+            auto addvertex_queue = bulk::queue<int, long, long>(world);
+            for (const auto& [sender, netid] : netidowner_queue) {
+                if (HC_list[nc_par].is_local_net(netid)) {
+                    for(auto v : HC_list[nc_par].net(netid).vertices()) {
+                        addvertex_queue(sender).send(world.rank(), v, netid);
+                    }
+                }
+            }
+            world.sync();
+
+            // Add pins locally from the queues
+            for (const auto& [sender, vtoadd, targetnet] : addvertex_queue) {
+                if (HC_list[nc_par].is_local_net(targetnet)) {
+                    HC_list[nc_par].net(targetnet).add_vertex(vtoadd);
+                }
+            }
 
-        // for (auto n = 0u; n < HC_list[nc_par].nets().size(); n++) {
-        //     if(ids[HC_list[nc_par].nets()[n].vertices()] != HC_list[nc_par].nets()[n].id()) {
-        //         HC_list[nc_par].nets()[n].set_cost(0);
-        //     }
-        //     else {
-        //         HC_list[nc_par].nets()[n].set_cost(edge_counts[HC_list[nc_par].nets()[n].vertices()]);
-        //     }
-        // }
+            // Initizalize maps for consensus building around hyperedge deletion
+            boost::unordered_map<std::vector<long>, long, VectorHasher> edge_counts;
+            std::unordered_map<long, long> comm_costs;
+            std::unordered_map<long, int> lowest_proc;
+
+            // Inefficient communication where costs are synchronized and 0 cost signifies deletion
+            // This passes tests and is sure to give correct results, more work is needed to get efficient
+            for (auto& n : HC_list[nc_par].nets()) {
+                if(edge_counts.find(n.vertices()) == edge_counts.end()) {

```

Comment:
I do not see why we need communication about the net costs. Whenever two nets are the same, they must also locally be the same. When we need to merge two nets, this is only relevant to processors that have vertices in these nets. If they have a vertex in one of the nets, they have the same vertex in the other. They thus already have the information on the cost of the nets that should be merged, they can just add these up using their local information. There only needs to be a deterministic way to select which net stays, for example simply the one with the lowest global id.

Reply:
I think that this was added to have an easy fix to coordinate which edge should be removed. While the local topology is not different, different processors may choose a different edge id to accumulate and communicate duplicate edges after this. But as I write this it dawns on me that the processors could simply locally choose the lowest number identifier and it would also work. So this is a major inefficiency I guess. Most of this code was to ensure correctness, not runtime.